### PR TITLE
Add: ScalarSpec to golden harness with .pt-uniform cache

### PR DESCRIPTION
**Pull request by zhangqi-chen**
## Summary

- New `golden/spec.py` unifies `TensorSpec` and `ScalarSpec`; old `golden/tensor_spec.py` removed
- `ScalarSpec` accepts `torch.dtype` mirroring `TensorSpec`; constructor takes `int`/`float`/`bool` or a 0-dim tensor and `__post_init__` coerces to a dtype-precise 0-dim `torch.Tensor` stored in `.value`
- Scalars persist as `data/in/{name}.pt` — same format used for tensor inputs, no separate JSON sidecar
- `run()` buckets all tensors before any scalar before passing args to `execute_compiled`, so users can interleave `Tensor`/`Scalar` freely in `specs` (works around simpler's `ChipStorageTaskArgs` ordering constraint without touching pypto or simpler)
- `run(... tensor_specs=)` keyword renamed to `run(... specs=)` and propagated across the 41 example call sites
- `examples/beginner/hello_world.py` reworked as the end-to-end smoke test for the scalar path: `ScalarSpec("a", torch.float32, 1.0)` interleaved between tensor input/output

## Related Issues

Related: hw-native-sys/pypto#1228 — runner-side bucketing here is the user-facing workaround for the runtime constraint described in that issue

### CODE REVIEW

File: golden/runner.py [hunk 1]
```diff
@@ -193,35 +204,59 @@ def _fail(error: str) -> RunResult:
         work_dir = compiled.output_dir
 
     # Generate Inputs
+    input_snapshot: dict[str, torch.Tensor] = {}
+    scalar_specs_eff: dict[str, ScalarSpec] = {}
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply golden/runner.py:208]: False positive — the `else:` at runner.py:289 is the else of `if data_dir is not None: ...`, so it only runs when `data_dir` is `None`. When `golden_data` is set, the cached `out/{name}.pt` is loaded instead of running `golden_fn`, per the docstring contract that `golden_data` takes precedence.

File: golden/spec.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        if self.dtype in _INT_TABLE:
+            return _INT_TABLE[self.dtype][0](int(self.value.item()))
+        if self.dtype is torch.bool:
+            return ctypes.c_bool(bool(self.value.item()))
+        if self.dtype is torch.float32:
+            return ctypes.c_float(float(self.value.item()))
+        # fp16 / bf16: read raw 16-bit pattern.
+        bits = self.value.view(torch.int16).item()
+        return ctypes.c_uint16(int(bits) & 0xFFFF)
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply golden/spec.py:208]: `int(bits) & 0xFFFF` correctly recovers the unsigned 16-bit pattern for `c_uint16`. `view(torch.uint16)` is unreliable across PyTorch versions (uint16 tensor support is only partial as of 2.3+). Keeping the int16+mask form.

File: golden/runner.py [hunk 3]
```diff
@@ -193,35 +204,59 @@ def _fail(error: str) -> RunResult:
...
+            # spec value, dtype must be identical).
+            for s in scalar_specs:
+                cached = torch.load(data_dir / "in" / f"{s.name}.pt", weights_only=True)
+                if cached.dtype != s.dtype:
+                    return _fail(
+                        f"{s.name}.pt dtype mismatch: spec={s.dtype} cache={cached.dtype}"
+                    )
+                scalar_specs_eff[s.name] = ScalarSpec(
+                    name=s.name, dtype=s.dtype, value=cached
+                )
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply golden/runner.py:247]: Fixed in 168d679 — added an explicit `cached.ndim != 0` (and non-tensor) check that returns via `_fail(...)` before constructing `ScalarSpec`. Regression test `test_scalar_pt_non_zero_dim_fails` added.

File: golden/spec.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            return self.init_value.to(dtype=self.dtype)
+        if callable(self.init_value):
+            # Support the standard torch factory functions used as callables
+            fn = self.init_value
+            if fn in (torch.randn, torch.rand, torch.zeros, torch.ones):
+                return fn(self.shape, dtype=self.dtype)
+            # Generic callable: call with no arguments, then cast
+            result: Any = fn()
+            return torch.as_tensor(result, dtype=self.dtype)
+        raise TypeError(f"Unsupported init_value type {type(self.init_value)!r} for tensor {self.name!r}")
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply golden/spec.py:83]: Out of scope for this PR — `TensorSpec.create_tensor` not enforcing `self.shape` is pre-existing behavior with an explicit regression pin in `test_tensor_init_ignores_spec_shape` ("regression pin; revisit if we decide to enforce shape matching"). Belongs in a follow-up.

File: tests/golden/test_runner.py [hunk 5]
```diff
@@ -52,25 +53,36 @@ def populated_cache(tmp_path):
...
-        if write_outputs_positional is None:
-            return
-        for tensor, value in zip(tensors, write_outputs_positional):
-            if value is not None:
-                tensor[:] = value
+    if fake_execute is None:
+        def fake_execute(work_dir, tensors, **kwargs):
+            if write_outputs_positional is None:
+                return
+            for tensor, value in zip(tensors, write_outputs_positional):
```
[the remark this answers is not in the document]
Comment (zhangqi-chen) [reply tests/golden/test_runner.py]: Fixed in 168d679 — added `strict=True` to the `zip(...)` call (also clears ruff B905).
